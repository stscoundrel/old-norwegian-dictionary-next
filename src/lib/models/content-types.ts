export enum ContentType {
    Word = 'word',
    Letter = 'letter',
    Page = 'page',
    Breadcrumbs = 'breadcrumbs',
    Other = 'other'
}

export default ContentType
