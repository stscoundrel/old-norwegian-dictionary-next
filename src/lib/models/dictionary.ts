import { DictionaryEntry as RawDictionaryEntry } from 'old-norwegian-dictionary/dist'

export type OriginalDictionaryEntry = RawDictionaryEntry

export interface DictionaryEntry extends RawDictionaryEntry {
    slug: string,
}
