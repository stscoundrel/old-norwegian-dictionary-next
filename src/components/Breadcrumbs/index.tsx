import { ContentType } from 'lib/models/content-types'
import { getBreadcrumbs } from 'lib/utils/breadcrumbs'
import { getSchema } from 'lib/utils/schema'
import Link from 'next/link'
import { DictionaryEntry } from 'lib/models/dictionary'
import styles from './Breadcrumbs.module.scss'

interface BreadcrumbProps{
  type: ContentType,
  content: DictionaryEntry | DictionaryEntry[] | null
}

export default function Breadcrumbs({ type, content }: BreadcrumbProps) {
  const getBreadCrumbData = () => {
    let letter: string | null = null
    let word: string | null = null

    if (content) {
      if (type !== ContentType.Page) {
        if (type === ContentType.Letter && Array.isArray(content)) {
          letter = content[0].word.charAt(0).toLowerCase()
        }

        if (type !== ContentType.Letter && !Array.isArray(content)) {
          letter = content.word.charAt(0).toLowerCase()
        }
      }

      if (type === ContentType.Word && !Array.isArray(content)) {
        word = content.word
      }
    }

    return {
      type,
      letter,
      word,
    }
  }

  const breadcrumbs = getBreadcrumbs(getBreadCrumbData())
  const schema = getSchema(breadcrumbs, ContentType.Breadcrumbs)

  return (
    <nav className={styles.section}>
      <div className="container">
        {breadcrumbs.map(({ label, url }) => (
          <Link key={url} href={url}>
            <a className={styles.link}>{label}</a>
          </Link>
        ))}

        <script type='application/ld+json' dangerouslySetInnerHTML={ { __html: schema } }/>
      </div>
    </nav>
  )
}
