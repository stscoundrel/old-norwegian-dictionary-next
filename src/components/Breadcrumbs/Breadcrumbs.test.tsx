import { ContentType } from 'lib/models/content-types'
import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import Breadcrumbs from './index'

describe('Breadcrumbs component', () => {
  const word = {
    word: 'úfrjáls',
    partOfSpeech: 'adj',
    definition: 'úfrjáls, adj.  1)  ufri, underkastet andresHerredømme som en Træl; oft verðr yðrúdrjúg hin úfrjálsa ættin til dreng-skaparins Flat. I, 33521.  2)  som manuhindret af Heftelse eller Paatale kanbeholde og benytte; om Jordegods: DN.II, 263. 252.',
    slug: 'ufrjals',
  }

  test('Default page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs type={ContentType.Page} content={null} />)
  })

  test('Default page: matches snapshot', () => {
    const tree = renderer.create(<Breadcrumbs type={ContentType.Page} content={null} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Default page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs type={ContentType.Page} content={null} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(1)
  })

  test('Letter page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs type={ContentType.Letter} content={[word]} />)
  })

  test('Letter page: matches snapshot', () => {
    const tree = renderer.create(
      <Breadcrumbs type={ContentType.Letter} content={[word]} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Letter page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs type={ContentType.Letter} content={[word]} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(2)
  })

  test('Word page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs type={ContentType.Word} content={word} />)
  })

  test('Word page: matches snapshot', () => {
    const tree = renderer.create(<Breadcrumbs type={ContentType.Word} content={word} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Word page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs type={ContentType.Word} content={word} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(3)
  })
})
