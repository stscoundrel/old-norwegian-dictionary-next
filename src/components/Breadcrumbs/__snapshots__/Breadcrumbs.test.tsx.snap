// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs component Default page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Norwegian Dictionary
    </a>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Norwegian Dictionary","item":"undefined/"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;

exports[`Breadcrumbs component Letter page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Norwegian Dictionary
    </a>
    <a
      className="link"
      href="/letter/u2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Letter Ú
    </a>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Norwegian Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter Ú","item":"undefined/letter/u2"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;

exports[`Breadcrumbs component Word page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Norwegian Dictionary
    </a>
    <a
      className="link"
      href="/letter/u2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Letter Ú
    </a>
    <a
      className="link"
      href="/word/ufrjals"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Úfrjáls
    </a>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Norwegian Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter Ú","item":"undefined/letter/u2"},{"@type":"ListItem","position":3,"name":"Úfrjáls","item":"undefined/word/ufrjals"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;
